File: 1. Even Or Odd & Swap XOR & Count Set Bits.py
# Topic: Check even/odd using n&1, swap using XOR, count set bits

def is_odd(n):
    return (n & 1) == 1

def swap_xor(a, b):
    a = a ^ b
    b = a ^ b
    a = a ^ b
    return a, b

def count_set_bits(n):
    cnt = 0
    while n:
        n &= (n-1)
        cnt += 1
    return cnt

if __name__ == '__main__':
    n = 13
    print(n, 'is odd' if is_odd(n) else 'is even')
    a, b = 5, 9
    print('Before swap:', a, b)
    a, b = swap_xor(a,b)
    print('After swap:', a, b)
    print('Set bits in', n, ':', count_set_bits(n))

# Time Complexity: O(number of set bits)
# Space Complexity: O(1)

File: 1. Even Or Odd & Swap XOR & Count Set Bits.cpp
// Topic: Check even/odd using n&1, swap using XOR, count set bits
#include <bits/stdc++.h>
using namespace std;

int countSetBits(int n)
{
    int cnt = 0;
    while (n)
    {
        n &= (n - 1);
        cnt++;
    }
    return cnt;
}

int main()
{
    int n = 13;  // 1101
    cout << n << ((n & 1) ? " is odd\n" : " is even\n");

    int a = 5, b = 9;  // swap using XOR
    cout << "Before swap: " << a << " " << b << '\n';
    a = a ^ b;
    b = a ^ b;
    a = a ^ b;
    cout << "After swap: " << a << " " << b << '\n';

    cout << "Set bits in " << n << ": " << countSetBits(n) << '\n';
    return 0;
}

// Time Complexity: O(number of set bits)
// Space Complexity: O(1)
